#20 implement category repository

MaksimKosyhin opened the pull request.

Commit:
Added Category Repository, improved configs, changed db structure
Additionally added component for image loading, added tests that aren't working due to configuration problems.

Commit:
Refactor config, replace h2 database for testing with testcontainers

MaksimKosyhin committed:
moved .gitignore to outer directory

MaksimKosyhin committed:
Delete .idea directory

Commit:
moved .gitignore to outer directory

MaksimKosyhin commented:
I've tried to use testcontainers for repository unit testing but application throws an exception like this: https://stackoverflow.com/questions/72795700/testcontainer-rabbitmq-seems-to-release-connection-as-soon-as-it-is-start

MaksimKosyhin commented:
I solved this problem. It didn't launch because of wrong property reference. That exception still appears but tests work.

igor-bratko commented on Marketplace/src/main/java/com/example/demo/category/JdbcCategoryRepository.java:
```diff
@@ -0,0 +1,108 @@
+        jdbcTemplate.update(removeProductShops);
+    }
+
+    //TODO: move this functionality into another repository and service
```
Never leave commented code

igor-bratko commented on Marketplace/src/main/java/com/example/demo/config/DataSourceConfig.java:
```diff
@@ -3,28 +3,49 @@
+		return new JdbcTemplate(dataSource);
+	}
+
+	@Bean DbProperties dbProperties() {
```
Remove this bean. You already have DbProperties bean with @ConfigurationProperties.  Simply inject it to dataSource() method

Commit:
Wrote category repository with test, finished configuring project.
Still need to write tests for getProducts() method.
Also need to refactor this method because query() method doesn't support adding values for db's IN operator

Commit:
Wrote category repository with test, finished configuring project.
Still need to write tests for getProducts() method.
Also need to refactor this method because query() method doesn't support adding values for db's IN operator

Commit:
Merge remote-tracking branch 'origin/#20-Implement-Category-Repository' into Implement-Category-Repository

Commit:
fix

Commit:
Merge remote-tracking branch 'origin/#20-Implement-Category-Repository' into #20-Implement-Category-Repository

MaksimKosyhin commented:
forgot to make a small change before commit so I tried to ammend it. But something went wrong with merging. Next commits are done well so I guess this is no longer a problem

Commit:
Updated getProducts() method of repository

igor-bratko commented on Marketplace/src/main/java/com/example/demo/category/JdbcCategoryRepository.java:
```diff
@@ -0,0 +1,122 @@
+
+    @Override
+    public List<ProductInfo> getProducts(ProductQuery productQuery) {
+        String queryShops =
```
Make a single sql. Usually sql doesn't splitted in this way

igor-bratko commented on Marketplace/src/main/java/com/example/demo/category/OrderBy.java:
```diff
@@ -0,0 +1,19 @@
+package com.example.demo.category;
+
+public enum OrderBy {
+    SCORE_ASC("ORDER BY average_score ASC"),
```
It's better leave only ASC,DESC. You can add additional field to ProductQuery with order field. And remember, that all sql related code must be in repository layer. Services and Rest Api shouldn't be aware about 'ORDER BY total_reviews DESC'. It's better to add mapping for sql related info to repository layer. If you have enum with order field SCORE,PRICE etc - add mapping to repository for db columns SCORE -> average_score etc

igor-bratko commented on Marketplace/src/main/java/com/example/demo/category/ProductQuery.java:
```diff
@@ -0,0 +1,38 @@
+
+public class ProductQuery {
+    private long categoryId;
+    private List<Long> characteristics;
```
Can you use long instead of Long?

igor-bratko commented on Marketplace/src/main/java/com/example/demo/category/ProductQuery.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public String getCharacteristicsInsertParametersTemplate() {
+        return String.join(",", Collections.nCopies(characteristics.size(), "?"));
```
Make convertation to sql in repository layer. You can create separate converter from query to sql

igor-bratko commented on Marketplace/src/main/java/com/example/demo/category/JdbcCategoryRepository.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    @Override
+    public void removeCategory(long categoryId) {
```
This method should be in transaction

igor-bratko commented on Marketplace/src/main/java/com/example/demo/category/Category.java:
```diff
@@ -0,0 +1,16 @@
+package com.example.demo.category;
```
rename root package com.example.demo -> com.marketplace

Commit:
Updated repository

Commit:
Updated repository

Commit:
Merge remote-tracking branch 'origin/#20-Implement-Category-Repository' into #20-Implement-Category-Repository

Commit:
Updated repository

MaksimKosyhin commented:
I've understood the commit problem before. it turns out I can't ammend already pushed commit

Commit:
Updated repository, refactor tests
